[NCCL] Use cudaEventQuery to Poll for GPU operation errors

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44163 [NCCL] Add Environment Variable to guard Async Error Handling feature
* #41054 [NCCL] Destructor Blocks on WorkNCCL Completion
* #41053 [NCCL] WorkNCCL Helper Functions
* #41052 [NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread
* **#41051 [NCCL] Use cudaEventQuery to Poll for GPU operation errors**
* #41050 [NCCL] Timeout Loop Thread for Async Error Handling

**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

osalpekar committed:
[NCCL] Register Guard Function Callback in WorkNCCL Synchronize

Creating host guard function that throws exception in error cases and registering guard function as a callback on current CUDA stream

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Register Guard Function Callback in WorkNCCL Synchronize"


Creating host guard function that throws exception in error cases and registering guard function as a callback on current CUDA stream

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Register Guard Function Callback in WorkNCCL Synchronize"


Creating host guard function that throws exception in error cases and registering guard function as a callback on current CUDA stream

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -307,8 +321,19 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
+    // Enqueue guard function as a callback on the current stream to throw
+    // user-stream exception upon error.
+    if (!blockingWait_) {
+      cudaStreamAddCallback(currentStream, errorGuard, this, 0);
```
IIRC @sinannasir mentioned that this callback leads to 30% perf regression due to H2D comm. Is that correct? cc @sinannasir 

Shall we run some benchmark to make sure this does not introduce the same regression here?

osalpekar replied:
@mrshenli We actually remove the `cudaStreamAddCallback` in the last diff in this stack and replace it with a `cudaEventQuery`-based solution (the stream callback was the original solution, and I simply added the change to the top of the stack to iterate on our experiments more quickly). I am working on refactoring this stack a bit so that we don't have this stream callback at all.

Regarding benchmarks, we've actually thoroughly benchmarked the final solution on TSC/CU workloads, and I will share a doc summarizing these findings. In summary, this entire stack introduces less than 1% QPS regression on TSC (with NE parity) and almost no QPS regression on CU.

osalpekar committed:
Update on "[NCCL] Register Guard Function Callback in WorkNCCL Synchronize"


Creating host guard function that throws exception in error cases and registering guard function as a callback on current CUDA stream

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -566,7 +574,11 @@ void ProcessGroupNCCL::workCleanupLoop() {
-        // Remove all Completed WorkNCCL Objects from the Vector
+        // Handle Exceptions on failed GPU operations and remove completed
+        // workNCCL objects from work vector.
+        if (work->finishedGPUExecution()) {
```
Why do we need to condition this on `work->finishedGPUExecution`?

osalpekar replied:
Removing this in latest version

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -566,7 +574,11 @@ void ProcessGroupNCCL::workCleanupLoop() {
+        // Handle Exceptions on failed GPU operations and remove completed
+        // workNCCL objects from work vector.
+        if (work->finishedGPUExecution()) {
+          work->handleNCCLGuard();
```
Throwing an exception here is going to kill this thread and also cause a process crash. This doesn't seem like a good idea in general. What is the reason to throw the exception from this thread itself?

osalpekar replied:
Referring to Hongyi's comment (https://github.com/pytorch/pytorch/pull/43232#discussion_r482569599). The premise of this work is prevent stuck/hanging GPU ops from from stalling the entire training process by throwing an exception (crashing the process), and then allowing something like torchelastic to recover and continue training from last checkpoint. We need a thread that periodically checks for exceptions (either due to errors or timeouts) in unfinished collectives and throws these exceptions, which we do in the workCleanupThread.

pritamdamania87 replied:
In case of errors or timeouts we actually abort the NCCL communicators. In that case subsequent NCCL calls will fail and an error would be raised to the user. Is this sufficient for what we're looking to do here?

My concern with this approach is that it drastically changes the default behavior for users today and many users don't actually use torchelastic. It would be pretty bad user experience if by default the process crashes on any NCCL errors/timeouts.

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Use cudaEventQuery to Poll for GPU operation errors"


**This Commit:**
In the workCleanupThread, we process completion and exception handling for workNCCL objects corresponding to collective calls that have either completed GPU Execution, or have already thrown an exception. This way, we throw an exception from the workCleanupThread for failed GPU operations. This approach replaces the previous (and lower performance) approach of enqueuing a callback on the CUDA stream to process failures.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21938498](https://our.internmc.facebook.com/intern/diff/D21938498/)

[ghstack-poisoned]
